File: src/gpu.c
#include <stdint.h>

#include "gpu.h"
#include "mem.h"

const uint16_t COLORS[4] = {0x0eef, 0x099a, 0x0445, 0x0001};
uint32_t clock = 0;
uint8_t mode = 0;

/* stubbed with code from here while I figure this out
 * http://imrannazar.com/GameBoy-Emulation-in-JavaScript:-GPU-Timings
 */
uint8_t gpu_step(uint16_t cycles, uint16_t *buffer) {
	uint8_t redraw = 0;
	clock += cycles;

	switch(mode) {
	// OAM read mode, scanline active
	case 2:
	if(clock >= 80) {
		// Enter scanline mode 3
		clock = 0;
		mode = 3;
		REG_STAT = (REG_STAT & 0xfc) | mode;
	}
	break;

	// VRAM read mode, scanline active
	// Treat end of mode 3 as end of scanline
	case 3:
	if(clock >= 172) {
		// Enter hblank
		clock = 0;
		mode = 0;
		REG_STAT = (REG_STAT & 0xfc) | mode;

		// Write a scanline to the framebuffer
		gpu_draw_scanline();
	}
	break;

	// Hblank
	// After the last hblank, push the screen data to canvas
	case 0:
	if(clock >= 204) {
		clock = 0;
		REG_LY++;

		if(REG_LY == 143) {
			// Enter vblank
			mode = 1;
			REG_STAT = (REG_STAT & 0xfc) | mode;
			//gpu_draw_screen(buffer);
			redraw = 1;
		} else {
			mode = 2;
			REG_STAT = (REG_STAT & 0xfc) | mode;
		}
	}
	break;

	// Vblank (10 lines)
	case 1:
	if(clock >= 456) {
		clock = 0;
		REG_LY++;

		if(REG_LY > 153) {
			// Restart scanning modes
			REG_LY = 0;
			mode = 2;
			REG_STAT = (REG_STAT & 0xfc) | mode;
		}
	}
	break;
	}

	return redraw;
}

void gpu_set_pixel(uint16_t *pixels, uint8_t x, uint8_t y, uint16_t color) {
	uint16_t sx = x & 0xff, sy = y & 0xff;
	pixels[(sy * 256) + sx] = color;
	//pixels[(y << 8) | x] = color;
}

void gpu_draw_tile(uint16_t src_addr, uint16_t *dst_buffer, uint16_t x, uint16_t y) {
	uint32_t r, c, color_index, line1, line2;
	for (r = 0; r < 8; r++) {
		line1 = mem_read8(src_addr++);
		line2 = mem_read8(src_addr++);
		for (c = 0; c < 8; c++) {
			color_index  = (line1 >> (7 - c)) & 1;
			color_index |= line2 & (0x80 >> c) ? 2 : 0;
			gpu_set_pixel(dst_buffer, x+c, y+r, COLORS[color_index]);
		}
	}
}

void gpu_draw_sprite(uint16_t src_addr, uint16_t *dst_buffer, uint16_t x, uint16_t y) {
	uint32_t r, c, color_index, line1, line2;
	for (r = 0; r < 8; r++) {
		line1 = mem_read8(src_addr++);
		line2 = mem_read8(src_addr++);
		for (c = 0; c < 8; c++) {
			color_index  = (line1 >> (7 - c)) & 1;
			color_index |= line2 & (0x80 >> c) ? 2 : 0;
			if (color_index > 0) {
				gpu_set_pixel(dst_buffer, x+c, y+r, COLORS[color_index]);
			}
		}
	}
}

void gpu_draw_scanline() {
	/* TODO */
}

uint16_t get_tile_addr(uint8_t id) {
	int8_t sid = (int8_t) id;
	if (REG_LCDC & LCDC_BG_TILE_DATA) {
		return (id * 16) + 0x8000;
	} else {
		return (sid * 16) + 0x9000;
	}
}

void gpu_draw_screen(uint16_t *buffer) {
	/* for now just draw all the tiles */
	uint16_t r, c, tile_id, tile_addr, tile_ptr = 0x9800;

	if (REG_LCDC & LCDC_BG_TILE_MAP_SELECT) {
		tile_ptr = 0x9c00;
	}

	for (r = 0; r < 32; r++) {
		for (c = 0; c < 32; c++) {
			tile_id = mem_read8(tile_ptr++);
			tile_addr = get_tile_addr(tile_id);
			gpu_draw_tile(tile_addr, buffer, c * 8 - REG_SCX, r * 8 - REG_SCY);
		}
	}

	uint8_t x, y, id, flags;
	uint16_t sprite_addr = 0xfe00;
	if (REG_LCDC & LCDC_SHOW_SPRITES) {
		for (r = 0; r < 40; r++) {
			x = mem_read8(sprite_addr++);
			y = mem_read8(sprite_addr++);
			id = mem_read8(sprite_addr++);
			flags = mem_read8(sprite_addr++);

			if (x == 0 && y == 0) continue;
			gpu_draw_sprite(id * 16 + 0x8000, buffer, x - REG_SCX, y - REG_SCY);

			if (REG_LCDC & LCDC_SPRITE_DOUBLE_HEIGHT) {
				gpu_draw_sprite(id * 16 + 0x8000 + 16, buffer, x - REG_SCX, y + 8 - REG_SCY);
			}
		}
	}
}

File: src/gpu.rs
use crate::memory::Memory;
use std::convert::TryInto;

// static LCDC_ON: u8 = 1 << 7;
static LCDC_WINDOW_TILE_MAP_SELECT: u8 = 1 << 6;
static LCDC_WINDOW_ON: u8 = 1 << 5;
static LCDC_BG_TILE_DATA: u8 = 1 << 4;
static LCDC_BG_TILE_MAP_SELECT: u8 = 1 << 3;
static LCDC_SPRITE_DOUBLE_HEIGHT: u8 = 1 << 2;
static LCDC_SHOW_SPRITES: u8 = 1 << 1;
// static LCDC_SHOW_BG: u8 = 1 << 0;


// static SPRITE_PRIORITY: u8 = 1 << 7;
static SPRITE_FLIP_V:u8 = 1 << 6;
static SPRITE_FLIP_H: u8 = 1 << 5;

static COLORS: [(u8, u8); 4] = [(0xe7, 0x9c), (0x97, 0x08), (0x44, 0x31), (0x31, 0x6a)];

pub struct GPU {
    pub clock: u16,
    pub mode: u8,
}

impl GPU {
    pub fn step(&mut self, mem: &mut Memory, cycles: u16) -> (bool, bool) {
        let mut redraw = false;
        let mut vblank = false;
        self.clock += cycles;

        match self.mode {
            0 => {
                // HBlank
                if self.clock >= 204 {
                    self.clock = 0;
                    *mem.reg_ly() = mem.reg_ly().wrapping_add(1);

                    if *mem.reg_ly() == 143 {
                        self.mode = 1;
                        *mem.reg_stat() = (*mem.reg_stat() & 0xfc) | self.mode;
                        vblank = true;
                        redraw = true;
                    } else {
                        self.mode = 2;
                        *mem.reg_stat() = (*mem.reg_stat() & 0xfc) | self.mode;
                    }
                }
            }
            1 => {
                // VBlank
                if self.clock >= 456 {
                    self.clock = 0;
                    *mem.reg_ly() = mem.reg_ly().wrapping_add(1);

                    if *mem.reg_ly() > 153 {
                        *mem.reg_ly() = 0;
                        self.mode = 2;
                        *mem.reg_stat() = (*mem.reg_stat() & 0xfc) | self.mode;
                    }
                }
            }
            2 => {
                // OAM read mode
                if self.clock >= 80 {
                    self.clock = 0;
                    self.mode = 3;
                    *mem.reg_stat() = (*mem.reg_stat() & 0xfc) | self.mode;
                }
            }
            3 => {
                // VRAM read mode
                if self.clock >= 172 {
                    self.clock = 0;
                    self.mode = 0;
                    *mem.reg_stat() = (*mem.reg_stat() & 0xfc) | self.mode;
                    // self.draw_scanline();
                }
            }
            _ => {
                panic!("Unhandled GPU mode {}", self.mode)
            }
        }
        (redraw, vblank)
    }

    #[inline]
    fn set_pixel(buffer: &mut [u8], x: u8, y: u8, color_index: usize) {
        let sx = x as usize;
        let sy = y as usize;
        let offset = ((sy * 256) + sx) * 2;
        let (top, bottom) = COLORS[color_index];
        buffer[offset] = bottom;
        buffer[offset+1] = top;
    }

    fn draw_tile(&self, mem: &Memory, tile_addr: u16, buffer: &mut [u8], x: u8, y: u8) {
        let mut mut_tile_addr = tile_addr;
        for r in 0..8 {
            let line1 = mem.read8(mut_tile_addr);
            mut_tile_addr += 1;
            let line2 = mem.read8(mut_tile_addr);
            mut_tile_addr += 1;
            for c in 0..8 {
                let mut color_index = (line1 >> (7 - c)) & 1;
                color_index |= if line2 & (0x80 >> c) != 0 { 2 } else { 0 };
                GPU::set_pixel(buffer, x.wrapping_add(c), y.wrapping_add(r), color_index as usize);
            }
        }
    }

    fn draw_sprite(&self, mem: &Memory, src_addr: u16, buffer: &mut [u8], x: u8, y: u8, flags: u8) {
        let mut mut_src_addr = src_addr;
        for r in 0..8 {
            let line1 = mem.read8(mut_src_addr);
            mut_src_addr += 1;
            let line2 = mem.read8(mut_src_addr);
            mut_src_addr += 1;
            for c in 0..8 {
                let mut color_index = (line1 >> (7 - c)) & 1;
                color_index |= if line2 & (0x80 >> c) != 0 { 2 } else { 0 };
                if color_index > 0 {
                    let mut gx = x.wrapping_add(c);
                    let mut gy = y.wrapping_add(r);
                    if (flags & SPRITE_FLIP_H) != 0 { gx = x.wrapping_add(7-c); }
                    if (flags & SPRITE_FLIP_V) != 0 { gy = y.wrapping_add(7-r); }
                    GPU::set_pixel(buffer, gx, gy, color_index as usize);
                }
            }
        }
    }

    fn get_tile_addr(&self, mem: &mut Memory, tile_id: u8) -> u16 {
        if *mem.reg_lcdc() & LCDC_BG_TILE_DATA != 0 {
            (tile_id as u16) * 16 + 0x8000
        } else {
            let tile_sid = (tile_id as i8) as i16 * 16;
            0x9000_u16.wrapping_add(tile_sid as u16)
        }
    }

    pub fn draw_screen(&self, mem: &mut Memory, buffer: &mut [u8]) {
        let mut tile_id: u8;
        let mut tile_addr: u16;
        let mut tile_ptr: u16 = 0x9800;

        if *mem.reg_lcdc() & LCDC_BG_TILE_MAP_SELECT != 0 {
            tile_ptr = 0x9c00;
        }

        // Tiles
        for r in 0..32 {
            for c in 0..32 {
                tile_id = mem.read8(tile_ptr);
                tile_ptr += 1;
                tile_addr = self.get_tile_addr(mem, tile_id);
                let scx = *mem.reg_scx();
                let scy = *mem.reg_scy();
                self.draw_tile(
                    mem,
                    tile_addr,
                    buffer,
                    ((c * 8u16).wrapping_sub(scx as u16) & 0xff).try_into().unwrap(),
                    ((r * 8u16).wrapping_sub(scy as u16) & 0xff).try_into().unwrap(),
                );
            }
        }

        let mut win_ptr: u16 = 0x9800;

        if *mem.reg_lcdc() & LCDC_WINDOW_TILE_MAP_SELECT != 0 {
            win_ptr = 0x9c00;
        }

        // Window
        if *mem.reg_lcdc() & LCDC_WINDOW_ON != 0 {
            for r in 0..32 {
                for c in 0..32 {
                    tile_id = mem.read8(win_ptr);
                    win_ptr += 1;
                    tile_addr = self.get_tile_addr(mem, tile_id);
                    let x = (c * 8u8).wrapping_add(*mem.reg_wx()).wrapping_sub(7);
                    let y = (r * 8u8).wrapping_add(*mem.reg_wy());
                    if tile_id != 0 && x < 167 && y < 144 {
                        self.draw_tile(mem, tile_addr, buffer, x, y);
                    }
                }
            }
        }

        // Sprites
        let mut id: u16;
        let mut flags: u8;
        let mut sprite_addr: u16 = 0xfe00;
        if *mem.reg_lcdc() & LCDC_SHOW_SPRITES != 0 {
            for _r in 0..40 {
                let y = mem.read8(sprite_addr).wrapping_sub(16);
                sprite_addr += 1;
                let x = mem.read8(sprite_addr).wrapping_sub(8);
                sprite_addr += 1;
                id = mem.read8(sprite_addr) as u16;
                sprite_addr += 1;
                flags = mem.read8(sprite_addr);
                sprite_addr += 1;

                if x == 0 && y == 0 {
                    continue;
                }
                self.draw_sprite(mem, id * 16 + 0x8000, buffer, x, y, flags);

                if *mem.reg_lcdc() & LCDC_SPRITE_DOUBLE_HEIGHT != 0 {
                    self.draw_sprite(mem, id * 16 + 0x8000 + 16, buffer, x, y + 8, flags);
                }
            }
        }
    }
}
